packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx

```diff
@@ -118,17 +118,19 @@ export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => {
   const getNearestIndexToRender = React.useCallback(
     (offset: number) => {
       const lastMeasuredIndexRelativeToAllRows = apiRef.current.unstable_getLastMeasuredRowIndex();
-      const lastMeasuredIndexRelativeToCurrentPage =
-        lastMeasuredIndexRelativeToAllRows - (currentPage.range?.firstRowIndex || 0);
-      const lastMeasuredIndex = Math.max(0, lastMeasuredIndexRelativeToCurrentPage);
-
-      let allRowsMeasured = lastMeasuredIndex === Infinity;
+      let allRowsMeasured = lastMeasuredIndexRelativeToAllRows === Infinity;
       if (currentPage.range?.lastRowIndex && !allRowsMeasured) {
         // Check if all rows in this page are already measured
-        allRowsMeasured = lastMeasuredIndex >= currentPage.range.lastRowIndex;
+        allRowsMeasured = lastMeasuredIndexRelativeToAllRows >= currentPage.range.lastRowIndex;
       }
 
-      if (allRowsMeasured || rowsMeta.positions[lastMeasuredIndex] >= offset) {
+      const lastMeasuredIndexRelativeToCurrentPage = clamp(
+        lastMeasuredIndexRelativeToAllRows - (currentPage.range?.firstRowIndex || 0),
+        0,
+        rowsMeta.positions.length,
+      );
+
+      if (allRowsMeasured || rowsMeta.positions[lastMeasuredIndexRelativeToCurrentPage] >= offset) {
```

Comment:
The bug in #6029 was caused because the index used to check with `rowsMeta.positions` was outside the bounds, so it would satisfy this condition and call `exponentialSearch` below. My fix was to adjust so the index used to check `rowsMeta.positions` is relative to the current page only.